### PR TITLE
Fix XmlSerializer config

### PR DESCRIPTION
**Pull request by timbussmann**
fixes https://github.com/Particular/NServiceBus/issues/3852

Turns out that all xml serializer config settings don't work.

Todo:
* [x] Acceptance tests for the remaining XmlSerializer config extensions
* [x] Fixing the remaining XmlSerializer config extensions
* [x] Remove `SetProperty` on the SettingsHolder class

@Particular/nservicebus-maintainers while it shouldn't be a big deal to fix this, I'm a bit worried about the impact on ASQ in case they use the `XmlSerializer` and are these config extensions. The settings are currently shared on all instances of the serializer, although that probably shouldn't have an impact? Even `DontWrapRawXml` shouldn't be an issue as long as the wrapper message doesn't use XContainer properties, right?

### CODE REVIEW

**Comment by danielmarbach** [comment]
@timbussmann I didn't fully understand your statement above. What are you worried about? Are you worried about removing `SetProperty`?

**Comment by bording** [comment]
@danielmarbach I believe his concern was more about the fact that these settings haven't been working at all, so would the header serialization that ASQ does break in some way with any of these settings actually applied.

**Comment by timbussmann** [comment]
@danielmarbach what @bording said :point_up:

**Comment by danielmarbach** [comment]
@timbussmann how can we answer those questions then?

**Comment by timbussmann** [comment]
I'd say testing it manually and pulling in @Particular/azure-storage-queues-maintainers

**Comment by timbussmann** [comment]
@Particular/azure-storage-queues-maintainers can you see any potential conflicts when these configurations (sanitize input, dontwraprawxml and custom namespace) are applied to the messagewrapper serializer?

@Particular/nservicebus-maintainers besides required testing, this should be ready for review

File: src/NServiceBus.AcceptanceTests/Serialization/When_configuring_custom_xml_namespace.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    using System.Threading.Tasks;
+    using System.Xml.Linq;
+    using AcceptanceTesting;
+    using EndpointTemplates;
+    using MessageMutator;
+    using NUnit.Framework;
+
+    public class When_configuring_custom_xml_namespace : NServiceBusAcceptanceTest
+    {
+        const string CustomXmlNamespace = "http://particular.net";
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Serialization/When_configuring_custom_xml_namespace.cs]: Just by looking at this constant I would have assumed that this is the default namespace :)

File: src/NServiceBus.AcceptanceTests/Serialization/When_sanitizing_xml_messages.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    using NUnit.Framework;
+
+    public class When_sanitizing_xml_messages : NServiceBusAcceptanceTest
+    {
+        [Test]
+        public async Task Should_remove_illegal_characters_from_messages()
+        {
+            var context = await Scenario.Define<Context>()
+                .WithEndpoint<EndpointSanitizingInput>(e => e
+                    .When(session => session.SendLocal(new SimpleMessage {Value = "Hello World!"})))
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Serialization/When_sanitizing_xml_messages.cs]: Shouldn't there be spacing like `new SimpleMessage { Value = "Hello World!" }`?

File: src/NServiceBus.AcceptanceTests/Serialization/When_sanitizing_xml_messages.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            {
+                public IncomingMutator(Context scenarioContext)
+                {
+                    this.scenarioContext = scenarioContext;
+                }
+
+                public Task MutateIncoming(MutateIncomingTransportMessageContext context)
+                {
+                    var body = Encoding.UTF8.GetString(context.Body);
+                    var invalidChar = char.ConvertFromUtf32(0x8);
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Serialization/When_sanitizing_xml_messages.cs]: Would it make sense to add a comment what kind of illegal char this is? For better understandability

File: src/NServiceBus.Core/obsoletes.cs [hunk 4]
```diff
@@ -2254,6 +2255,24 @@ public void ApplyTo(Type componentType, IComponentConfig config)
         {
             throw new NotImplementedException();
         }
+
+        [ObsoleteEx(
+          RemoveInVersion = "7.0",
+          TreatAsErrorFromVersion = "6.0",
+          Message = "Set(string key, object value)")]
```
Comment (danielmarbach) [inline src/NServiceBus.Core/obsoletes.cs]: I think when you do it like you did you have to use ReplacementTypeOrMember

But I don't think this is enough. I would appreciate a more thorough explanation like

Setting properties on an object is no longer supported. Set the value in the settings dictionary by using Set(string key, object value) and pass the settings holder to the object needing the settings. Extract the settings and set the state accordingly.

Sorry just typed this out of the void but something similar.

File: src/NServiceBus.Core/obsoletes.cs [hunk 5]
```diff
@@ -2254,6 +2255,24 @@ public void ApplyTo(Type componentType, IComponentConfig config)
...
+        public void SetProperty<T>(Expression<Func<T, object>> property, object value)
+        {
+            throw new NotImplementedException();
+        }
+
+        [ObsoleteEx(
+          RemoveInVersion = "7.0",
+          TreatAsErrorFromVersion = "6.0",
+          Message = "Set(string key, object value)")]
```
Comment (danielmarbach) [inline src/NServiceBus.Core/obsoletes.cs]: See above

File: src/NServiceBus.AcceptanceTests/Serialization/When_skip_wrapping_xml.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+    using System.Xml.Linq;
+    using AcceptanceTesting;
+    using EndpointTemplates;
+    using MessageMutator;
+    using NUnit.Framework;
+
+    public class When_skip_wrapping_xml : NServiceBusAcceptanceTest
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Serialization/When_skip_wrapping_xml.cs]: Do we need a test similar to this which does not skip to show the difference?

File: src/NServiceBus.AcceptanceTests/NServiceBus.AcceptanceTests.csproj [hunk 7]
```diff
@@ -133,7 +134,7 @@
     <Compile Include="Performance\When_message_is_audited.cs" />
     <Compile Include="Basic\When_receiving_with_catch_all_handlers_registered.cs" />
     <Compile Include="Basic\When_handling_message_with_several_messagehandlers.cs" />
-    <Compile Include="Basic\When_registering_custom_serializer.cs" />
+    <Compile Include="Serialization\When_registering_custom_serializer.cs" />
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/NServiceBus.AcceptanceTests.csproj]: Should we exclude those tests from the source package for downstream?

**Comment by danielmarbach** [comment]
@timbussmann made a few comments

File: src/NServiceBus.AcceptanceTests/Serialization/When_sanitizing_xml_messages.cs (hunk 3 above)
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Serialization/When_sanitizing_xml_messages.cs]: I have no idea what character that is tbh. I just took one which was on the filtered chars. Apparently it's not very easy to sneak in one of the chars above 0x0Fxxxx(whatever that range was) since these can't be converted to chars (while 0x8 can). Otherwise we would need to replace raw bytes.

File: src/NServiceBus.Core/obsoletes.cs (hunk 4 above)
Comment (timbussmann) [inline src/NServiceBus.Core/obsoletes.cs]: yeah, good catch, need to use ReplacementTypeOrMember :+1: 

I'm on the fence about the obsolete message. I doubt that anyone is using it (because it doesn't nearly work as you'd expect it).
> and pass the settings holder to the object needing the settings. Extract the settings and set the state accordingly.

that was already needed before. The SetProperty did nothing else than just a Set with the object full name + propertyname. It doesn't change anything on the access side.

Comment (danielmarbach) [inline src/NServiceBus.Core/obsoletes.cs]: See your argument. I'm good then

File: src/NServiceBus.AcceptanceTests/Serialization/When_skip_wrapping_xml.cs (hunk 6 above)
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Serialization/When_skip_wrapping_xml.cs]: so a test testing that the xml containing invalid characters fails deserialization and moves to the error queue? not sure, is that worth it since it's a violation of the xml spec?

Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Serialization/When_skip_wrapping_xml.cs]: No I meant

When_wrapping_is_not_skipped.Should_wrap_xml_content

Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Serialization/When_skip_wrapping_xml.cs]: urgh, sorry I messed the tests up in my head. Yeah that could probably make sense.

File: src/NServiceBus.AcceptanceTests/NServiceBus.AcceptanceTests.csproj (hunk 7 above)
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/NServiceBus.AcceptanceTests.csproj]: I think this is out of scope?

**Comment by timbussmann** [comment]
@danielmarbach added `When_wrapping_is_not_skipped.Should_wrap_xml_content` as an acceptance test

File: src/NServiceBus.AcceptanceTests/NServiceBus.AcceptanceTests.csproj (hunk 7 above)
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/NServiceBus.AcceptanceTests.csproj]: btw. I think these tests would have caught the XmlSerializer issues with ASB

**Comment by timbussmann** [comment]
btw. checked with ASQ and the wrapper does use a `byte[]` for the content and no XDocument (actually obvious since you can use a different serializer for the inner message), so there should be no issue with `DontWrapRawXml`.

**Comment by bording** [comment]
:+1: LGTM

**Comment by danielmarbach** [comment]
Build is failing. Btw. I don't think exluding these tests is out of scope. Since we introduce new ATs we should at least think if they are required downstream and if not exclude it

**Comment by timbussmann** [comment]
looks like a TC issue to me

@danielmarbach I remain with my statement

> btw. I think these tests would have caught the XmlSerializer issues with ASB

I think it adds value to run the serializers on downstreams, we would have found bugs much earlier on with these test.

**Comment by danielmarbach** [comment]
@timbussmann fair enough
